Dynamic style update for feature meshes

Neptilo opened the pull request:
Addresses #2649

## Description

Enables dynamic style updates for feature meshes. Refactors buffer logic into reusable functions and adds style versioning to track property changes.

## Motivation and Context

Allows real-time style updates (colors, heights) without rebuilding meshes, improving performance for interactive applications.

Neptilo committed:
feat: dynamically update mesh colors

Neptilo committed:
feat: dynamically update extruded polygon style properties

Neptilo committed:
fix: correctly compare colors

Neptilo committed:
feat: implement line extrusion as cylinders

Neptilo committed:
perf: don't update feature style that was just initialized

Neptilo committed:
feat: implement updating line style

Neptilo committed:
feat: implement style updates for remaining feature types

Neptilo committed:
perf: optimize extruded line construction

Neptilo committed:
feat: implement basic joints between cylinders

Neptilo committed:
refactor: write vertex and index buffers in same loop for extruded lines

Neptilo committed:
feat: generate spherical wedges at extruded line joints

Neptilo committed:
refactor: update style method to accept collection context in Feature2Mesh

Neptilo committed:
chore: trim everything related to extruded lines

Neptilo committed:
chore: clean up comments

Neptilo committed:
feat: store original matrix world for collection and update during context setting

Neptilo committed:
fix: opacity calculation in ReferLayerProperties function

Neptilo committed:
refactor: don't recreate color buffer attribute object on update

Neptilo committed:
fix: tests

Neptilo committed:
refactor: remove unused code

ftoromanoff commented:
Thanks for the PR ! 

Would it be possible to add unit tests for the new behavior that you add here ? (Mostly for Feature2Mesh.js)
<img width="762" height="202" alt="image" src="https://github.com/user-attachments/assets/d5664e60-8195-4c52-9ee8-1d90b1a3e34c" />

And maybe an example where we can play with this  ?

Neptilo committed:
chore: add tests for updating feature styles

Neptilo committed:
feat: add buttons to toggle some styles in WFS example

Neptilo committed:
feat: add unit test for updating child mesh styles in FeatureProcessing

Neptilo commented:
@ftoromanoff @mgermerie PR ready again

jailln commented:
ping @ftoromanoff @mgermerie

ftoromanoff commented:
I don't know if it's linked with that PR, but when I play with the Layer WFS Building' opacity and that I tilt the view I get some strange results:
<img width="1330" height="875" alt="image" src="https://github.com/user-attachments/assets/cc711a69-9fff-4bfc-a101-19d1a55a2e5d" />
(This doesn't seem to appear in ortho view)

ftoromanoff commented on packages/Main/src/Converter/Feature2Mesh.js:
```diff
@@ -702,4 +871,69 @@ export default {
+        collection.matrixWorld.copy(collection.origMatrixWorld);
+
+        // only define attributes that need an update
+        featureMesh.stylePropVersions ||= {};
```
I don't really understand the use of the propertie 'stylePropVersions'.
If it's just to see if it need an update, why not just use a 'needUpdate' as boolean ?
Moreover, I fell like that if we switch way and back the toggling, we will create a new stylePropVersions each time even if it will be the same style... (kind of a memory leak?)

Neptilo replied:
It was a long time ago so I'll do my best to remember...
If we use a simple boolean, the idea would be that we reset it as soon as the new style property has been "consumed" by an object, but in this case we have many objects consuming the style change, so we don't know when it is finished.
The next idea was to compare the values between the source style and the target objects, and only update them if they have changed, but we have the problem that the style value may be defined as a callback, but the getter doesn't return that function but tries to call it.
Hence the next idea of storing a property version number. Note that this is a common technique also used in Three internals.
If we keep toggling the switch, the property version will increase each time indeed, but since it is just a number, it will not use more memory.

ftoromanoff replied:
I feel it strange to add a 'style' property in featureMesh when there is already a style object exactly for that..

Neptilo replied:
Are you still referring to `stylePropVersions`? That only contains numbers, not styles per se. Would you like to discuss this orally so we can move forward?

Neptilo replied:
After discussion, we decided to trigger style updates from the setter of the style properties (done by the iTowns user) instead of doing it on the layer updates, using an event mechanism instead of a property version mechanism.

Neptilo replied:
For information, the refactoring works for **WFS** datasets but broke the style updates for **MVT** dataset (e.g. `vector_tile_3d_mesh`). I've tried debugging it but haven't been able to find a solution so far.

mgermerie commented on packages/Main/src/Core/Style.js:
```diff
@@ -192,6 +193,10 @@ export class StyleContext {
+    getGeometry() {
+        return this.#geometry;
+    }
+
```
Maybe a proper `get geometry` would be more alike what's done with `properties` getter. I know setters are not defined as so but we can try at least to have coherence in our getters.

mgermerie commented on packages/Main/src/Converter/Feature2Mesh.js:
```diff
@@ -514,24 +620,87 @@ function featureToExtrudedPolygon(feature, options) {
+ * @param {number} buffers.indexPtr - Current write position in index buffer (incremented by function).
+ * @param {number} [id] - Batch id value to assign to written vertices when batchIds is provided.
+ */
+function updateLineBuffers(featureMesh, buffers, id) {
```
For more code navigability and consistency with your other updates, can you either define this method right after `featureToLine` or define all other `update{shape}Buffer` methods here along side this one ?

mgermerie commented on packages/Main/src/Converter/Feature2Mesh.js:
```diff
@@ -187,47 +195,24 @@ function featureToPoint(feature, options) {
+    up.set(0, 0, 1).multiply(inverseScale);
 
     const pointMaterialSize = [];
     context.setFeature(feature);
```
You re-call this inside the `updatePointBuffer` method. I get the need of it inside `updatePointBuffer` method (which can be called by `updateStyle`) but it thus should be removed from here. Same goes for `context.setGeometry` on line 205 and in other feature type methods.

Neptilo replied:
Good catch. For `context.setGeometry`, I don't see the redundancy. All the `update*Buffers` seem to access `context.geometry` and therefore need the geometry to be defined before.

mgermerie replied:
Reading the code again, there is indeed no issue with `context.geometry`. What initially concerned me is that `context.geometry` is always set by the method calling `update{Shape}Buffers`, but is only used within `update{Shape}Buffers`. This could lead to side effects if `update{Shape}Buffers` is called improperly (i.e., without setting `context.geometry` first). However, I don’t see a clean way to avoid this without looping through each geometry inside `update{Style}Method`, which wouldn’t be ideal.

mgermerie reviewed (changes requested):
Thanks for the PR, and nice idea for the style properties versioning implementation!

A few comments on the code other than which there is an issue with line altitude: it seems you have an accidental increment somewhere, making each line appear higher than the previous one. You can see this behavior on this screenshot:

<img width="1920" height="1080" alt="Image" src="https://github.com/user-attachments/assets/1bfe2b8e-9b0e-46a1-8fb3-6bb412ac9468" />

mgermerie commented:
> I don't know if it's linked with that PR, but when I play with the Layer WFS Building' opacity and that I tilt the view I get some strange results: <img alt="image" width="1330" height="875" src="https://private-user-images.githubusercontent.com/47628509/546116682-cc711a69-9fff-4bfc-a101-19d1a55a2e5d.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzAzOTEyODksIm5iZiI6MTc3MDM5MDk4OSwicGF0aCI6Ii80NzYyODUwOS81NDYxMTY2ODItY2M3MTFhNjktOWZmZi00YmZjLWExMDEtMTlkMWE1NWEyZTVkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNjAyMDYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjYwMjA2VDE1MTYyOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTAzZDE4ZjViNGNlZWNiMzcxZTA2NzkwODQ2NmJjZjQyYzhlNDc1YjFlMjE3NzFlNzFlOWM1MDc0Yzc2MTc3ZDUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.hsmFY-MT_4-8Unfx96Q2wL5jr305wav5xHjw51-y46U"> (This doesn't seem to appear in ortho view)

From my personal testing, I did not have this issue, buildings opacity was correctly computed. I tried on the [WFS to 3D objects](https://www.itowns-project.org/itowns/examples/source_stream_wfs_3d.html) example, with custom or identical color for each buildings, before and after toggling their extrusion height. Maybe you found this in another test condition @ftoromanoff ?

Neptilo committed:
fix: get base_altitude inside for loop

Neptilo commented:
> A few comments on the code other than which there is an issue with line altitude: it seems you have an accidental increment somewhere, making each line appear higher than the previous one.

Found it! I had oversimplified by getting `base_altitude`  before the for loop, but actually, the call to `setLocalCoordinatesFromArray` inside the loop could modify the context, which was then used in `base_altitude`'s getter.

Neptilo commented:
> I don't know if it's linked with that PR, but when I play with the Layer WFS Building' opacity and that I tilt the view I get some strange results

I can´t reproduce either. In the image it looks like the terrain also has transparency.

Neptilo committed:
refactor: improve code

ftoromanoff commented:
> > I don't know if it's linked with that PR, but when I play with the Layer WFS Building' opacity and that I tilt the view I get some strange results
> 
> I can´t reproduce either. In the image it looks like the terrain also has transparency.

I tested again and don't have it anymore.... Might have been a wrong version on my side...
I will just advice, for a better understanding in the example to have the 'Toggle Extrusion' and 'Toggle Color' after the Layer WFS Building (or in the folder), you can also migrate that exemple to lil-gui as I did for the PointCloud examples (way back^^)

Neptilo committed:
refactor: migrate WFS example's GUI integration with lil-gui

ftoromanoff reviewed (commented):
Sorry, it seems that i forgot to submit this commentary...

gchoqueux commented on examples/source_stream_wfs_3d.html:
```diff
@@ -218,25 +226,33 @@
+            var debugGui = new lil();
+
+            // Add buttons to toggle extrusion height and color
+            const extrusionToggle = {
+                toggleHeight: function () {
+                    useFixedHeight = !useFixedHeight;
+                    wfsBuildingLayer.style.fill.extrusion_height = useFixedHeight ? 8 : extrudeBuildings
+                    view.notifyChange(wfsBuildingLayer);
+                },
+                toggleColor: function () {
+                    useFixedColor = !useFixedColor;
+                    wfsBuildingLayer.style.fill.color = useFixedColor ? color.set(0x00cc00) : colorBuildings
+                    view.notifyChange(wfsBuildingLayer);
+                }
+            };
+
```
Can you use the callbacks (`colorBuildings, extrudeBuildings`) already present in the example to specify colors and height based on properties and context?

Neptilo replied:
Not sure what you mean. Those callbacks are already used.

gchoqueux replied:
Could you modify  the callbacks `colorBuildings` and  `extrudeBuildings`  to specify by example : 

```js
function colorBuildings(properties) {
    if (toogle) { 
         return '#00ff00';
    }
    if (properties.usage_1 === 'Résidentiel') {
        return '#FDFDFF';
    } else if (properties.usage_1 === 'Annexe') {
        return '#C6C5B9';
    } else if (properties.usage_1 === 'Commercial et services') {
        return '#62929E';
    } else if (properties.usage_1 === 'Religieux') {
        return '#393D3F';
    } else if (properties.usage_1 === 'Sportif') {
        return '#546A7B';
    }

    return '#555555';
}
```

Neptilo replied:
In that case, that would technically be the same style, not a style change, and the toggle wouldn't trigger an update.

ftoromanoff replied:
But do we really need a 'style change' ?

Neptilo replied:
As I said, without it, the toggle wouldn't trigger an update, so the objects would keep the same appearance as before. And we can´t reasonably defer the reponsibility of calling `style.dispatchEvent({ type: 'style-property-changed', style, parameter });` to the user. Even with a wrapper around it, the user would need to pass the style and changed parameter name as arguments (to keep the benefit of not updating all buffers).
And even if we implemented such a user-called update signal, as a user I would still exepect `style.property = ...` to work out-of-the-box. So the current example is useful to at least demonstrate that this technique works.

gchoqueux commented on packages/Main/src/Layer/ReferencingLayerProperties.js:
```diff
@@ -4,18 +4,21 @@ function ReferLayerProperties(material, layer) {
     if (layer && layer.isGeometryLayer) {
         material.layer = layer;
 
+        const getOpacity = (opacity) => {
+            const styleOpacity = material.layer.style?.fill?.opacity ?? 1;
+            const layerOpacity = material.layer.opacity;
+            return styleOpacity * layerOpacity * (opacity ?? 1);
+        };
```
`opacity` and `styleOpacity` are different ?

Neptilo replied:
Updated. They are different but they are never both defined at the same time.

gchoqueux reviewed (dismissed):
## Architecture

### Pattern
A robust way possible is use  [Model–view–controller](https://en.wikipedia.org/wiki/Model%E2%80%93view%E2%80%93controller) to apply dynamic style applies a style to data.
Could you reinforce this pattern within your development?

### Update Style 

When the building height modification is applied, it seems to me that all buildings are rebuilt: reconstruction of the base and roof vertices, face computation, and repopulation of the colors. Is that correct?  If so, what would be the difference compared to removing the `featureMesh` and recreating it ?

####  Proposed solutions

1. Create a controller/updater to modify only the relevant attribute
2.  Link `batchId` attributes to style buffer attributes. You would only need to modify the attributes. wdyt ?
3. use [BatchedMesh](https://threejs.org/docs/?q=batch#BatchedMesh) see [example ](https://threejs.org/examples/?q=bat#webgl_batch_lod_bvh)

## Example

Could add a color picker to specify the color building ?
Could you add a slider to change height building in real  time ? 
Is it possible to have a style property that changes over time — that is, one that can be animated?

## Issue

There's a issue with one color mode, it's not just the color green

<img width="1782" height="928" alt="Image" src="https://github.com/user-attachments/assets/be4ee947-dcac-42b2-b3f7-b54af51dbda3" />

Neptilo committed:
fix: colors in example

Neptilo committed:
chore: simplify getOpacity function

ftoromanoff commented on packages/Main/src/Process/FeatureProcessing.js:
```diff
@@ -25,6 +26,11 @@ export default {
+
+                // update existing features
+                for (const featureMesh of f.meshes.children) {
+                    Feature2Mesh.updateStyle(featureMesh, f.collection, layer.style);
```
Why not a featureMesh.applyStyle(), as this don't update a style but a featureMesh.

The style actually if i recall corectly, fixed a value (for each style property) at start between the value entered by the user (the layer.value) then the value from the source, then the default value. Maybe we just need an other level, or a way to have the userValue modified, As i can see ther is a set property in defineStyleProperty than can have a function. Can you have a function that use the value from your toggle button ?

Neptilo replied:
I think the second comment is the same as https://github.com/iTowns/itowns/pull/2648#discussion_r2864344681. See the discussions there.

ftoromanoff replied:
Indeed, it's linked.

Neptilo committed:
feat: use event to apply style property change

instead of comparing property versions and checking for updates on each layer update

Neptilo committed:
fix: update tests after `updateStyles` refactoring

Neptilo committed:
fix: evaluate color per extruded polygon vertex to fix WFS 2.5D example

ftoromanoff commented on examples/source_stream_wfs_3d.html:
```diff
@@ -269,24 +285,38 @@
+                        var gui = debug.GeometryDebug.createGeometryDebugUI(debugGui, view, layer);
                         debug.GeometryDebug.addWireFrameCheckbox(gui, view, layer);
+
+                        // Buttons to demonstrate dynamic style updates
```
As it's not really a 'dynamic style' update, I would suggest to change the commentary to 'Buttons to demonstrate a dynamic update of style.' or 'Buttons to demonstrate  the feasibilty of a change of style after initialization.'

ftoromanoff commented on packages/Main/src/Converter/Feature2Mesh.js:
```diff
@@ -13,7 +13,7 @@ let style;
 const dim_ref = new THREE.Vector2();
 const dim = new THREE.Vector2();
-const normal = new THREE.Vector3();
+const up = new THREE.Vector3();
```
Maybe we should go back to 'normal' or find an other expliciter name ?

ftoromanoff replied:
up is indeed a property on THREE.object3D that "Defines the up direction of the 3D object which influences the orientation".
But is it the same property here ? just to be sure we don't get confused...

Neptilo replied:
This is indeed the idea I had in mind.
In practice, `up` is mostly used in Three for objects on which it is meaningful to define the action of "looking at" something: cameras, lights, or even sprites (that must keep facing the camera).
Here, `up` is used as the direction of extrusion, so not the same use case, but it does represent the same vector: the `up` of the extruded polygons represents the same vector as the `up` of the camera.

ftoromanoff commented on packages/Main/src/Converter/Feature2Mesh.js:
```diff
@@ -38,17 +38,25 @@ class FeatureMesh extends THREE.Group {
         this.#collection.scale.copy(collection.scale);
         this.#collection.updateMatrix();
+        this.#collection.matrixWorldInverse = this.#collection.matrixWorld.clone().invert();
+        this.#collection.origMatrixWorld = collection.matrixWorld.clone();
```
Why do you need to save the matrixWorld ?

Neptilo replied:
To reapply it here: https://github.com/iTowns/itowns/pull/2648#discussion_r2980118683.
I can't really explain it, but without it the result was incorrect (before the event refactoring - and after the refactoring it is more broken).

ftoromanoff commented on packages/Main/src/Converter/Feature2Mesh.js:
```diff
@@ -38,17 +38,25 @@ class FeatureMesh extends THREE.Group {
         this.#originalCrs = collection.crs;
         this.#currentCrs = this.#originalCrs;
         this.extent = collection.extent;
+        this.isFeatureMesh = true;
```
Do you need that property ?

Neptilo replied:
It is useful when debugging, to check if an object is of this type. It was already done in some classes (and everywhere within Three's codebase) so it seemed reasonable to keep it here too.

ftoromanoff commented on packages/Main/src/Converter/Feature2Mesh.js:
```diff
@@ -698,8 +860,78 @@ export default {
+
+    context.setCollection(collection);
+    collection.updateMatrixWorld(true);
+    collection.matrixWorld.copy(collection.origMatrixWorld);
```
You update the MatrixWorld to overwrite the next line. I feel there is somethings weird here...

Neptilo replied:
See https://github.com/iTowns/itowns/pull/2648#discussion_r2996411147.
The results were incorrect if we removed one of those lines. A subtlety about `updateMatrixWorld` is that it also updates its children's matrices.

mgermerie commented on examples/source_stream_wfs_3d.html:
```diff
@@ -218,25 +226,33 @@
+            const extrusionToggle = {
+                toggleHeight: function () {
+                    useFixedHeight = !useFixedHeight;
+                    wfsBuildingLayer.style.fill.extrusion_height = useFixedHeight ? 8 : extrudeBuildings
```
Missed it on the first review, but could you just add the `;` at the end of this instruction as well as on line 244 ?

mgermerie reviewed (dismissed):
Thanks for the architecture updates on the PR!

The current approach to updating styles does not account for changes in line width, which I missed during my initial review. Was this an oversight, or is it out of scope for this PR?

If it’s the latter, could you please:
- add a console warning to inform users that changing `style.stroke.width` is not yet supported?
- open an issue to track support for this feature?

Neptilo committed:
fix: fix style update for MVT datasets

Neptilo committed:
fix: use matrix instead of matrixWorld

jailln commented:
@mgermerie @ftoromanoff  I took a look into the two remaining topics of this PR:

1. `collection` `matrixWorld` that was saved in `origMatrixWorld` in `convert` and reused in `applyStyle`. 

It was related to an issue that was already present but not visible before: We used `collection.matrixWorld` instead of `collection.matrix` - which was not an issue in `convert` since they were the same but is an issue in `applyStyle` since the collection has been added to the scene at that point. This is fixed in [the following commit](https://github.com/iTowns/itowns/pull/2648/changes/2e9b80bd268bd1a09fd87fa5fc3c33c3760d9a2d)

2. The need to store a styleVersion in each Layer node (instead of having a layer.needsUpdate property on the layer only). 

When not using style versions in each layer node, there was a bug where some tiles didn't get the correct style modifications. I found out that this was the case for cached tiles. When the user triggered a style modification, all displayed tiles were correctly updated and and then layer.needsUpdate was reset to false (but cached tiles style was not updated). Then, when a cached tile was retrieved to be displayed, it contained the wrong style and did not went through `applyStyle` because layer.needsUpdate was already false. Applying style updates to all cached tiles whenever a user changes a style property would be a waste of time since cached tiles may not ever be displayed. However I tried an implementation without the layer node style versioning system (see [this commit](https://github.com/iTowns/itowns/commit/047d097f615503d71da52ae43c2d8a1e9e3281b2)). The main idea was to apply the style modification to tiles in the scheduler callback ([here](https://github.com/iTowns/itowns/commit/047d097f615503d71da52ae43c2d8a1e9e3281b2#diff-073c93060804b9a6a67da1185aeb54d5f9613ae80abe7e84834563e2d7c688efR89-R100)). However, this also applies it to newly loaded tiles, whether they are retrieved from the cache or not, which is not optimized at all (newly loaded tiles buffers are always computed twice). Therefore, we think that the best implementation is to have a style versionning mechanism on tiles to know if they need to be updated or not to correctly account for cached tiles, but to also use events to mark layers that needs update (which is an optimization comparing to the first proposed implementation). This is implemented in the last commits pushed on the branch of this PR.

We are available to explain this orally since it may be difficult to follow in written.

Neptilo committed:
fix: make minor changes in example

Desplandis reviewed (changes requested):
From my understanding, this PR has two main goals:
- First, a refactoring of the `Feature2Mesh` module into smaller, composable components which I fully welcome. I didn't notice any issues here except in my opinion a lack of tests (was already the case on master)
- Second, a style property change feature and not a dynamic style per se (if you could update the PR title accordingly). I have a few concerns, primarily regarding the style-change API, which currently does not support some expected usage patterns (see the dedicated comment).

Regarding the change notification process, I understand why and event-based approach alone isn't sufficient, since buffer updates of a loaded feature mesh only occurs when it is updated by the layer (which may not occur if it was culled for this frame). However, a simpler mechanism (such as tracking nodes pending updates within the layer) might be easier to reason about, maintain and test. I don't see the need for versioning when a simple flag or tracked node set could do the job.

jailln committed:
feat(style): account for different style properties setting pattern

Especially style categories modification and layer style object modification.

jailln committed:
feat(style): fix tests for style property change
